Add Universe Opt In Source

localjo opened the pull request:
I'd like to add support for a Universe API parameter to subscribe-email that allows the source of an opt-in to be defined

https://github.com/SparkartGroupInc/sparkart-services/blob/master/doc/api/contacts.md

`contact[lists_opt_in_source]`

localjo committed:
Update README.md

Spec out the new function via the README. RDD FTW!

josiahsprague commented:
Something about adding arbitrary parameters for only one API doesn't settle right with me. I'm trying to think if there is a good way to wrap service-specific options so that it's clear that an option only works with service x.

localjo commented:
Thinking through some ways to approach this while being conscious of how other mailing list providers handle things so we don't end up with an unwieldy set of options that are different for each provider. What if we did something like this?

```
var mySubscribeForm = new SubscribeEmail({
  element: '#subscribe-form',
  service: 'universe',
  key: 'api-key-here',
  customFields: {
    'contact[lists_opt_in_source]': {
      type: 'hidden',
      value: 'Some string'
    },
    'contact[first_name]': {
      type: 'text',
      label: 'First Name'
    }
  }
});
```

That way we could account for any possible fields someone might want to add to their subscribe form. In Universe, the field options are pretty consistent. But in MailChimp and SendGrid, custom fields can be added for whatever an account manager might want to collect.

localjo commented:
Another way of looking at it is that this could be a good use case for custom templates. Just add a custom template with `<input type="hidden" name="contact[lists_opt_in_source]" value="Opt In Source Name">`. I actually like this option. But maybe writing some verbose documentation on how to achieve it would be useful. Thoughts @pushred ?

pushred commented:
Yeah you're right, custom fields is exactly how [MailChimp](http://blog.mailchimp.com/tracking-signup-pages-via-the-mailchimp-embedded-forms/) and [Campaign Monitor](http://help.campaignmonitor.com/topic.aspx?t=185) handle this for instance. Universe also supports custom fields but this standard parameter is better because we're tracking email imports using it as well.

There's some advantage to the `customFields` approach, in the case of this particular parameter it'd likely be populated from JS anyway — site domain, a path, page title. It's more direct to fetch that data at instantiation rather than after the template has been rendered. We have a JS template at our disposal, which is really just an orderly way of doing DOM manipulation at the end of the day. This also allows the default template to be used more often.

I think what I was after here was a way of baking this in such that the source is guaranteed to be submitted, since it's important for compliance and list hygiene. But if other services generally require additional setup anyway then I agree some documentation would be helpful.

josiahsprague commented:
How about something like;
```
hiddenFields: [
  name: 'contact[lists_opt_in_source]',
  value: 'Some string'
]
```

For adding data that should be submitted by the form but doesn't affect what's displayed on a page. I think it makes sense to add non-displayed elements, such as hidden inputs, without making a custom template, but if displayed elements need to be added, such as text inputs, a custom template should be used.

localjo committed:
[#add] add hiddenFields option including tests

Branch: add-opt-in-field

localjo commented:
Added an option to do;

```
hiddenFields: [
  {
    name: 'contact[lists_opt_in_source]',
    value: 'My Marketing Campaign'
  }
]
```

This can also be used for other hidden inputs in order to add data that should be submitted by the form but doesn't affect what's displayed on a page. I think it makes sense to add non-displayed elements, such as hidden inputs, without making a custom template, but if displayed elements need to be added, such as text inputs, a custom template should be used.

Also added tests and updated the README.

pushred commented on README.md:
```diff
@@ -62,6 +62,18 @@ var mySubscribeForm = new SubscribeEmail({
 **(Required for MailChimp)** A string of the `<form action="">` attribute generated by MailChimp which contains MailChimp authentication information. You can get this from MailChimp under "Signup forms > Embedded form code > Naked" and copying just the value from the `<form action="">` attribute. It should follow this format: `http://{username}.{data center}.list-manage.com/subscribe/post?u={user ...
 
+### `hiddenFields`
+An optional array of objects used to add `<input type="hidden">` elements to the default template to allow additional data to be submitted along with a form. As an example, on Universe forms this can be used to track an opt-in source using the [`lists_opt_in_source` parameter](https://github.com/SparkartGroupInc/sparkart-services/blob/master/doc/api/contacts.md);
```
Unfortunately can't link to the Universe docs at the moment, that's a private repo. Maybe reference the MailChimp example instead?

pushred commented on src/subscribe-form.hbs:
```diff
@@ -1,4 +1,7 @@
 <form id="{{id}}" data-subscribe-instance="{{instanceId}}" class="{{namespace}}">
+{{#each hiddenFields}}
+<input class="{{../namespace}}__input {{../namespace}}__input--hidden" type="hidden" name="{{name}}" value="{{value}}">
```
What's the value of CSS classes on hidden inputs?

localjo replied:
Well, in theory those classes could be used for JS too. I was using them as handles for the tests, though that's not really necessary. They're not really necessary here, just added for consistency with the rest of the elements.

pushred commented on test/tests.js:
```diff
@@ -154,4 +154,28 @@ describe('Subscribe Email Module', function() {
+      var testField1 = $('#test-element .subscribe-email__input--hidden[name="testField1"]').val('testValue1');
+      assert(testField1.length > 0);
+      var testField2 = $('#test-element .subscribe-email__input--hidden[name="testField2"]').val('testValue2');
+      assert(testField2.length > 0);
```
Confused about what this is testing.. why are you setting values on the fields? Shouldn't the assertion be checking for the values from the config?

localjo committed:
[#add] check test value instead of setting

Branch: add-opt-in-field

localjo committed:
[#add] update hiddenFields example

Branch: add-opt-in-field

localjo committed:
[#add] test that input type is hidden, remove class

make sure that the input type in the template is actually hidden, and remove unnececarry classes

Branch: add-opt-in-field

localjo commented:
@pushred I'm not sure if this is still needed, but I think it's ready to merge.

localjo commented:
Doing some housekeeping and closing this PR due to lack of activity. 
![giphy](https://user-images.githubusercontent.com/516388/44671051-62d36100-a9f2-11e8-92b5-daa36ce72151.gif)
